### PR TITLE
Second dimension support for binop

### PR DESCRIPTION
**Pull request by umadevimcw**
### Ticket
#13646 

### Problem description
Missing second dimension support for binary op

### What's changed
Currently used repeat function to achieve the functionality

### Checklist
- [x] [Post commit CI passes](https://github.com/tenstorrent/tt-metal/actions/runs/11273939090)

### CODE REVIEW

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp [hunk 1]
```diff
@@ -117,6 +117,12 @@ auto preprocess_inputs(
             Shape repeats(std::array<uint32_t, 4>{first_shape[0], 1, 1, 1});
             second = ttnn::repeat(second, repeats);
         }
+        // repeats second if it is smaller
+        if (first_shape.rank() == 4 and second_shape.rank() == 4 and first_shape[1] > second_shape[1]) {
+            tt::log_warning(tt::LogOp, "Using repeat op to broadcast channel dim");
+            Shape repeats(std::array<uint32_t, 4>{1, first_shape[1], 1, 1});
+            second = ttnn::repeat(second, repeats);
+        }
     };
```
Comment (eyonland) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp:126]: Can we not put this temporary fix here instead?  https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp#L110

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp [hunk 2]
```diff
@@ -117,6 +117,12 @@ auto preprocess_inputs(
             Shape repeats(std::array<uint32_t, 4>{first_shape[0], 1, 1, 1});
             second = ttnn::repeat(second, repeats);
         }
+        // repeats second if it is smaller
+        if (first_shape.rank() == 4 and second_shape.rank() == 4 and first_shape[1] > second_shape[1]) {
+            tt::log_warning(tt::LogOp, "Using repeat op to broadcast channel dim");
+            Shape repeats(std::array<uint32_t, 4>{1, first_shape[1], 1, 1});
```
Comment (eyonland) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp:123]: Let's not log this and slow the perf down even more for this temp fix.

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp [hunk 3]
```diff
@@ -117,6 +117,12 @@ auto preprocess_inputs(
             Shape repeats(std::array<uint32_t, 4>{first_shape[0], 1, 1, 1});
```
Comment (eyonland) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp:117]: We need to remove this as well.

**Review by eyonland — COMMENTED** [review]
Can you pleas add a unit test?

File: ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp [hunk 4]
```diff
@@ -113,10 +113,14 @@ auto preprocess_inputs(
 
         // repeats second if it is smaller
         if (first_shape.rank() == 4 and second_shape.rank() == 4 and first_shape[0] > second_shape[0]) {
-            tt::log_warning(tt::LogOp, "Using repeat op to broadcast batch dim");
             Shape repeats(std::array<uint32_t, 4>{first_shape[0], 1, 1, 1});
             second = ttnn::repeat(second, repeats);
         }
+        // repeats second if it is smaller
+        if (first_shape.rank() == 4 and second_shape.rank() == 4 and first_shape[1] > second_shape[1]) {
+            Shape repeats(std::array<uint32_t, 4>{1, first_shape[1], 1, 1});
```
Comment (rfurko-tt) [inline ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp:121]: Isn't it a bug? I think we should pass `first_shape[1]/second_shape[1]` (also check whether `first_shape[1]` divisible by `second_shape[1]`)

Same applies to repeat for batch dimension.

Comment (rfurko-tt) [reply ttnn/cpp/ttnn/operations/eltwise/binary/binary.cpp:121]: Seems like pytorch doesn't broadcast if another dimension is not 1. Maybe the proper solution is to throw exception if dimension we trying to broadcast is not equal to 1. 
Screenshot 2024-10-15 at 10 25 50 PM

**Review by Aswinmcw — APPROVED** [review]
LGTM

**Review by VirdhatchaniKN — APPROVED** [review]

**Review by eyonland — APPROVED** [review]
Approving this for now to unblock.  I'm aware we need to get a more long term solution where we broadcast directly within the op itself.

**Pull request merged**